File: sdk-tools/ts2hc/src/logEvent.ts

export enum LogEventType { diagnostic, progressTick, progressStart, progressStop, error, }

export interface DiagnosticLogEvent {
    readonly logEventType: LogEventType.diagnostic;
    readonly message: string;
}

export interface ErrorLogEvent {
    readonly logEventType: LogEventType.error;
    readonly message: string;

}

export interface ProgressStartLogEvent {
    readonly logEventType: LogEventType.progressStart;
    readonly progessId: string;
}

export interface ProgressStopLogEvent {
    readonly logEventType: LogEventType.progressStop;
    readonly progessId: string;
}

export interface ProgressTickLogEvent {
    readonly logEventType: LogEventType.progressTick;
    readonly currentStep: number;
    readonly totalSteps: number;
    readonly suffixMessage: string;
    readonly progessId: string;
}

File: lib/build-project/logEvent.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'logEvent.freezed.dart';
part 'logEvent.g.dart';

@freezed
class DiagnosticLogEvent with _$DiagnosticLogEvent {
  const factory DiagnosticLogEvent({
    required int logEventType,
    required String message,
  }) = _$DiagnosticLogEventCtor;

  factory DiagnosticLogEvent.fromJson(Map<String, dynamic> json) =>
      _$DiagnosticLogEventFromJson(json);
}

@freezed
class ErrorLogEvent with _$ErrorLogEvent {
  const factory ErrorLogEvent({
    required int logEventType,
    required String message,
  }) = _$ErrorLogEventCtor;

  factory ErrorLogEvent.fromJson(Map<String, dynamic> json) =>
      _$ErrorLogEventFromJson(json);
}

@freezed
class ProgressStartLogEvent with _$ProgressStartLogEvent {
  const factory ProgressStartLogEvent({
    required int logEventType,
    required String progressId,
  }) = _$ProgressStartLogEventCtor;

  factory ProgressStartLogEvent.fromJson(Map<String, dynamic> json) =>
      _$ProgressStartLogEventFromJson(json);
}

@freezed
class ProgressStopLogEvent with _$ProgressStopLogEvent {
  const factory ProgressStopLogEvent({
    required int logEventType,
    required String progressId,
  }) = _$ProgressStopLogEventCtor;

  factory ProgressStopLogEvent.fromJson(Map<String, dynamic> json) =>
      _$ProgressStopLogEventFromJson(json);
}

@freezed
class ProgressTickLogEvent with _$ProgressTickLogEvent {
  const factory ProgressTickLogEvent({
    required int logEventType,
    required int currentStep,
    required int totalSteps,
    required String suffixMessage,
    required String progressId,
  }) = _$ProgressTickLogEventCtor;

  factory ProgressTickLogEvent.fromJson(Map<String, dynamic> json) =>
      _$ProgressTickLogEventFromJson(json);
}
